File: src/modelbox/tool/flow.cc
/*
 * Copyright 2021 The Modelbox Project Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


#include "flow.h"

#include <modelbox/modelbox.h>
#include <modelbox/flow.h>
#include <getopt.h>
#include <stdio.h>

#include <fstream>
#include <iostream>
#include <nlohmann/json.hpp>

namespace modelbox {

REG_MODELBOX_TOOL_COMMAND(ToolCommandFlow)

enum MODELBOX_TOOL_FLOW_COMMAND {
  MODELBOX_TOOL_FLOW_RUN,
  MODELBOX_TOOL_FLOW_CONF_CONVERT,
};

enum MODELBOX_TOOL_FLOW_CONVERT_COMMAND {
  MODELBOX_TOOL_FLOW_CONVERT_COMMAND_PATH,
  MODELBOX_TOOL_FLOW_CONVERT_COMMAND_OUTFORMAT,
};

static struct option flow_convert_options[] = {
    {"path", 1, nullptr, MODELBOX_TOOL_FLOW_CONVERT_COMMAND_PATH},
    {"out-format", 1, nullptr, MODELBOX_TOOL_FLOW_CONVERT_COMMAND_OUTFORMAT},
    {nullptr, 0, nullptr, 0},
};
static struct option flow_options[] = {
    {"run", 1, nullptr, MODELBOX_TOOL_FLOW_RUN},
    {"conf-convert", 0, nullptr, MODELBOX_TOOL_FLOW_CONF_CONVERT},
    {nullptr, 0, nullptr, 0},
};

ToolCommandFlow::ToolCommandFlow() = default;
ToolCommandFlow::~ToolCommandFlow() = default;

std::string ToolCommandFlow::GetHelp() {
  char help[] =
      " option:\n"
      "   -run [toml file]          run flow from file\n"
      "   -conf-convert             convert graph file format to json or "
      "toml\n"
      "     -path [conf file]       graph file path\n"
      "     -out-format [json|toml| output format, default is toml\n"
      "\n";
  return help;
}

int ToolCommandFlow::Run(int argc, char *argv[]) {
  int cmdtype = 0;
  int ret = -1;

  if (argc == 1) {
    std::cerr << GetHelp();
    return 1;
  }

  MODELBOX_COMMAND_GETOPT_BEGIN(cmdtype, flow_options)
  switch (cmdtype) {
    case MODELBOX_TOOL_FLOW_RUN:
      return RunFlow(optarg);
    case MODELBOX_TOOL_FLOW_CONF_CONVERT:
      optind = 1;
      MODELBOX_COMMAND_SUB_UNLOCK();
      return RunConfConvertCommand(MODELBOX_COMMAND_SUB_ARGC,
                                   MODELBOX_COMMAND_SUB_ARGV);
    default:
      break;
  }
  MODELBOX_COMMAND_GETOPT_END()

  return ret;
}

int ToolCommandFlow::RunFlow(const std::string &file) {
  auto flow = std::make_shared<modelbox::Flow>();
  MBLOG_INFO << "run flow " << file;
  auto ret = flow->Init(file);
  if (!ret) {
    MBLOG_ERROR << "init flow failed, " << ret.WrapErrormsgs();
    return 1;
  }

  ret = flow->Build();
  if (!ret) {
    MBLOG_ERROR << "build flow failed, " << ret.WrapErrormsgs();
    return 1;
  }

  flow->RunAsync();

  ret = flow->Wait();
  if (!ret) {
    MBLOG_ERROR << "run flow failed, " << ret.WrapErrormsgs();
    return 1;
  }

  flow->Stop();
  MBLOG_INFO << "run flow " << file << " success";

  return 0;
}

int ToolCommandFlow::RunConfConvertCommand(int argc, char *argv[]) {
  int cmdtype = 0;
  ConfigurationBuilder config_builder;
  std::shared_ptr<Configuration> config_merge;
  std::string path;
  std::string format = "toml";
  std::string out_result;
  modelbox::Status ret;

  MODELBOX_COMMAND_GETOPT_BEGIN(cmdtype, flow_convert_options)
  switch (cmdtype) {
    case MODELBOX_TOOL_FLOW_CONVERT_COMMAND_PATH:
      path = optarg;
      break;
    case MODELBOX_TOOL_FLOW_CONVERT_COMMAND_OUTFORMAT:
      format = optarg;
      break;
    default:
      break;
  }
  MODELBOX_COMMAND_GETOPT_END()
  if (path.length() == 0 || format.length() == 0) {
    std::cerr << "please input conf file path and format." << std::endl;
    return 1;
  }

  std::ifstream infile(path);
  if (infile.fail()) {
    std::cerr << "read file " << path << " failed, " << modelbox::StrError(errno)
              << std::endl;
    return 1;
  }

  Defer { infile.close(); };
  std::string data((std::istreambuf_iterator<char>(infile)),
                   std::istreambuf_iterator<char>());

  if (format == "json") {
    ret = modelbox::TomlToJson(data, &out_result, true);
  } else if (format == "toml") {
    ret = modelbox::JsonToToml(data, &out_result);
  } else {
    std::cerr << "output format is not supported" << std::endl;
    return 1;
  }

  if (!ret) {
    std::cerr << "convert failed, " << ret.WrapErrormsgs() << std::endl;
    return 1;
  }

  std::cout << out_result << std::endl;

  return 0;
}  // namespace modelbox

}  // namespace modelbox

File: src/java/src/main/java/com/modelbox/Flow.java
/*
 * Copyright 2021 The Modelbox Project Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.modelbox;

/**
 * modelbox Flow
 */
public class Flow extends NativeObject {
  public Flow() {
    setNativeHandle(FlowNew());
  }

  /**
   * init flow from inline graph
   * @param name graph name
   * @param graph inline graph
   * @throws ModelBoxException
   */
  public void init(String name, String graph) throws ModelBoxException {
    FlowInit(name, graph);
  }

  /**
   * init flow from graph file
   * @param file path to graph file
   * @throws ModelBoxException
   */
  public void init(String file) throws ModelBoxException {
    FlowInit(file);
  }

  /**
   * init flow by name, args and flow directory
   * @param name flow name
   * @param args flow args
   * @param flowDir scan flow directory
   * @throws ModelBoxException
   */
  public void initByName(String name, Configuration args, String flowDir) throws ModelBoxException {
    FlowInitByName(name, args, flowDir);
  }

  /**
   * init flow by name, args and flow directory
   * @param name flow name
   * @param args flow args
   * @throws ModelBoxException
   */
  public void initByName(String name, Configuration args) throws ModelBoxException {
    FlowInitByName(name, args);
  }

  /**
   * init flow by name, args and flow directory
   * @param name flow name
   * @param args flow args
   * @throws ModelBoxException
   */
  public void initByName(String name) throws ModelBoxException {
    FlowInitByName(name, null);
  }

  /**
   * Start run flow
   * @throws ModelBoxException
   */
  public void startRun() throws ModelBoxException {
    FlowStartRun();
  }

  /**
   * Wait flow finish
   * @throws ModelBoxException
   */
  public void waitFor() throws ModelBoxException {
    waitFor(0);
  }

  /**
   * Wait flow finish
   * @param timeout wait timeout, in millisecond 
   * @return whether timeout
   * @throws ModelBoxException
   */
  public boolean waitFor(long timeout) throws ModelBoxException {
    Status retval = new Status();
    return waitFor(timeout, retval);
  }

  /**
   * Wait flor finish, and get flow result
   * @param timeout wait timeout, in millisecond 
   * @param retval flow result.
   * @return whether timeout
   * @throws ModelBoxException
   */
  public boolean waitFor(long timeout, Status retval) throws ModelBoxException {
    return FlowWait(timeout, retval);
  }

  /**
   * Stop flow
   * @throws ModelBoxException
   */
  public void stop() throws ModelBoxException {
    FlowStop();
  }

  /**
   * Create external data for sending data to flow
   * @return ExternalDataMap object
   * @throws ModelBoxException
   */
  public ExternalDataMap createExternalDataMap() throws ModelBoxException {
    return FlowCreateExternalDataMap();
  }

  /**
   * Create stream io to send and recv stream data
   * @return ExternalDataMap object
   * @throws ModelBoxException
   */
  public FlowStreamIO CreateStreamIO() throws ModelBoxException {
    return FlowCreateStreamIO();
  }

  private native long FlowNew();

  private native boolean FlowWait(long timeout, Status status);

  private native void FlowStartRun();

  private native void FlowInit(String name, String graph);

  private native void FlowInit(String file);

  private native void FlowInitByName(String name, Configuration args, String flowDir);

  private native void FlowInitByName(String name, Configuration args);

  private native void FlowStop();

  private native ExternalDataMap FlowCreateExternalDataMap();

  private native FlowStreamIO FlowCreateStreamIO();
}
